Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -129,11 +140,42 @@ extension HTTPClient {
                 }
             }
 
+            func socketPathFromURL(_ url: URL) throws -> String {
+                switch self {
+                case .unixSocket(let scheme):
+                    switch scheme {
+                    case .baseURL:
+                        return url.baseURL?.path ?? url.path
+                    case .http_unix, .https_unix:
+                        guard let socketPath = url.host else {
+                            throw HTTPClientError.missingSocketPath
+                        }
+                        return socketPath
+                    }
+                case .host:
+                    return ""
+                }
+            }
+            
+            func uriFromURL(_ url: URL) -> String {
+                switch self {
+                case .host:
+                    return url.uri
+                case .unixSocket(let scheme):
+                    switch scheme {
+                    case .baseURL:
+                        return url.baseURL != nil ? url.uri : "/"
+                    case .http_unix, .https_unix:
+                        return url.uri
+                    }
+                }
+            }
+
             func supports(scheme: String) -> Bool {
                 switch self {
                 case .host:
                     return Kind.hostSchemes.contains(scheme)
-                case .unixSocket:
+                case .unixSocket(scheme: _):
                     return Kind.unixSchemes.contains(scheme)
```

Comment:
This `supports` spelling is getting unwieldy here, not least because we're no longer using half of the `switch` statement. Should this just be a constructor of `Kind` from `String`?

Reply:
I'm not too sure — we are already constructing a kind from a String in https://github.com/swift-server/async-http-client/pull/228/files#diff-67e2f60fc709da32af89dfbfc0c774aaR116-R129, so that method can certainly be fixed up to no longer rely on `supports()`.

My bigger question is that `supports()` seems like it is being used by some redirect logic here: https://github.com/swift-server/async-http-client/blob/070c1e5f374c76d5f55c2b4b60cbb7aa04471301/Sources/AsyncHTTPClient/HTTPHandler.swift#L962

<img width="1301" alt="image" src="https://user-images.githubusercontent.com/225505/83038487-deac7400-9ff1-11ea-9efa-e4f22ab5a554.png">
It seems like the older version was just making sure that the new scheme matched http/https - while the one currently in the repo will only accept http redirects from http requests, and socket redirects from socket requests. @krzyzanowskim Do you remember more about why you set it up this way?

I propose that if the request is an http request, we keep it as such, but if it is a socket request, we allow it to redirect to any supported scheme, but that is based on my limited understanding of what's going on here 😅

Reply:
> I propose that if the request is an http request, we keep it as such, but if it is a socket request, we allow it to redirect to any supported scheme, but that is based on my limited understanding of what's going on here 😅

Yeah, that seems a reasonable design choice.

Reply:
Would you support refactoring this method to something like this, since it is now only used for redirects?
<img width="1293" alt="image" src="https://user-images.githubusercontent.com/225505/83043233-752f6400-9ff7-11ea-9f5a-aa71c90bf962.png">

Call site (which was checking against the old scheme for some reason that escapes me):
<img width="1292" alt="image" src="https://user-images.githubusercontent.com/225505/83043339-98f2aa00-9ff7-11ea-9e41-3506b9712e99.png">


This would also simplify the initializer tremendously:
<img width="1292" alt="image" src="https://user-images.githubusercontent.com/225505/83042751-db67b700-9ff6-11ea-8f11-7edf986d47fb.png">

Reply:
> @krzyzanowskim Do you remember more about why you set it up this way?

Sorry, I don't remember details at the moment. I'm sorry not being helpful here.

Reply:
@krzyzanowskim Not a problem! I think the original code may have also had the same issue, so ultimately we just need more tests that capture the underlying problem 😅